### PR TITLE
Added ElementResolver member to WorkspaceModel

### PR DESCRIPTION
**Pull request by aparajit-pratap**
This PR is in anticipation of a follow up submission: https://github.com/DynamoDS/Dynamo/pull/3255

A new class `ElementResolver` has been added (that is used in: https://github.com/DynamoDS/Dynamo/pull/3255). It is made a member of `WorkspaceModel` as it stores a namespace mapping for all classes used in CBN's belonging to the same workspace. The mapping info in the `ElementResolver` needs to be accessed by `CodeBlockNodeModel` during pre-compilation in order to read from and/or update the map. Therefore this instance is packaged in an object of type `UpdateValueParams`, a new class, that is passed into `UpdateModelValue` that in turn can be passed on to the `CodeBlockNodeModel` object.

The `ElementResolver` object will be instantiated during workspace loading from the XML file before being passed to the `WorkspaceModel` being instantiated. At the time of saving the workspace, it will be serialized to the XML (DYN) file. However for purposes of this PR, since the `ElementResolver` is not consumed yet, this serialization and de-serialization implementation is deferred for a subsequent PR.

@Benglin please review as discussed. Thanks.

### CODE REVIEW

File: src/DynamoCore/Models/ModelBase.cs [hunk 1]
```diff
@@ -201,15 +202,14 @@ public bool HandleModelEvent(string eventName, UndoRedoRecorder recorder)
...
         /// </summary>
-        /// <param name="name">The name of a property/value to update.</param>
-        /// <param name="value">The new value to be set. This value comes directly
+        /// <param name="updateValueParams">It comprises 1. the name of a property/value to update.
+        /// 2. The new value to be set. This value comes directly
         /// from DynamoTextBox after user commits it. Overridden methods then use 
         /// a specific IValueConverter to turn this string into another data type 
-        /// that it expects.</param>
-        /// <param name="recorder"></param>
+        /// that it expects, 3. element resolver</param>
```
Comment (Benglin) [inline src/DynamoCore/Models/ModelBase.cs]: Thanks for updating these comments, @aparajit-pratap! I'd request this to be further updated to:

```csharp
/// <param name="updateValueParams">Please see UpdateValueParams for details.</param>
```

File: src/DynamoCore/Models/CustomNodeWorkspaceModel.cs [hunk 2]
```diff
@@ -32,7 +33,7 @@ private set
 
         public CustomNodeWorkspaceModel(
             string name, string category, string description, double x, double y, Guid customNodeId,
-            NodeFactory factory, string fileName="")
+            NodeFactory factory, ElementResolver elementResolver, string fileName="")
```
Comment (Benglin) [inline src/DynamoCore/Models/CustomNodeWorkspaceModel.cs]: Can `WorkspaceModel` not create `ElementResolver` internally? If it can, then we should definitely remove this, and make a `private readonly ElementResolver elementResolver` that can only be instantiated in the base `WorkspaceModel`. Derived workspace classes can use `WorkspaceModel.ElementResolver` internal property to access it.

File: src/DynamoCore/Search/SearchDictionary.cs [hunk 3]
```diff
@@ -229,10 +229,7 @@ private static string SanitizeQuery(string query)
             return query.Trim()
                 .Replace("\\", "\\\\")
                 .Replace(".", "\\.")
-                .Replace("*", "\\*")
-                .Replace("(", "\\(")
-                .Replace(")", "\\)")
-                .Replace("[", "\\[");
+                .Replace("*", "\\*");
```
Comment (Benglin) [inline src/DynamoCore/Search/SearchDictionary.cs]: Reverting your previous fix?

File: src/DynamoCore/Utilities/UpdateValueParams.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
+﻿using ProtoCore.Namespace;
```
Comment (Benglin) [inline src/DynamoCore/Utilities/UpdateValueParams.cs]: Please move this class alongside `ModelBase` class because that's where `UpdateValue` method is introduced, don't see the need of having a new class (hey, two files down).

File: src/Engine/ProtoCore/Namespace/ElementResolver.cs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+using System.Text;
+
+namespace ProtoCore.Namespace
+{
+    /// <summary>
+    /// Responsible for resolving a partial class name to its fully resolved name
+    /// </summary>
+    public class ElementResolver
+    {
+        private Dictionary<string, KeyValuePair<string, string>> resolutionMap;
```
Comment (Benglin) [inline src/Engine/ProtoCore/Namespace/ElementResolver.cs]: Should be made `readonly` (by the way, this class shouldn't be part of this PR, maybe all we need is an empty class declaration. It sorta *pollutes* the PR now).

File: src/Engine/ProtoCore/Namespace/ElementResolver.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        }
+
+        #region public constructors and methods
+
+        public ElementResolver()
+        {
+            resolutionMap = new Dictionary<string, KeyValuePair<string, string>>();
+        }
+
+        public ElementResolver(Dictionary<string, KeyValuePair<string, string>> namespaceLookupMap)
```
Comment (Benglin) [inline src/Engine/ProtoCore/Namespace/ElementResolver.cs]: I think the equivalent of `IEnumerable` should be used here (something like `IDictionary`?)

File: src/DynamoCore/Utilities/UpdateValueParams.cs [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+namespace Dynamo.Utilities
+{
+    public class UpdateValueParams
+    {
+        public UpdateValueParams(string value, string propertyName, ElementResolver elementResolver = null)
```
Comment (Benglin) [inline src/DynamoCore/Utilities/UpdateValueParams.cs]: Uh oh, I think there's a slight problem. The exact order of the argument must be 'propertyName', followed by 'propertyValue' (though it does not make a difference but from the API design standpoint I'd need it to be that way):

```csharp
/// <summary>
/// Creates an instance of UpdateValueParams object.
/// </summary>
/// <param name="propertyName">Name of the property whose value is to be updated.
/// This parameter cannot be empty or null.</param>
/// <param name="propertyValue">Value of the named property whose value is to be 
/// updated. This parameter can either be null or empty if the targeted property 
/// allows such values.</param>
/// 
public UpdateValueParams(string propertyName, string propertyValue,
    ElementResolver elementResolver = null)
{
}
```

File: src/DynamoCore/Core/CustomNodeManager.cs [hunk 8]
```diff
@@ -489,7 +492,7 @@ public bool TryGetInfoFromPath(string path, bool isTestMode, out CustomNodeInfo
                 nodeGraph.Notes,
                 workspaceInfo.X,
                 workspaceInfo.Y,
-                functionId, workspaceInfo.FileName);
+                functionId, elementResolver, workspaceInfo.FileName);
```
Comment (aparajit-pratap) [inline src/DynamoCore/Core/CustomNodeManager.cs]: The reason I'm not instantiating `ElementResolver` property inside the `WorkspaceModel` ctor is because it needs to be passed in from outside such as in this case. In my subsequent PR, `ElementResolver` will be obtained from `LoadGraphFromXml` with a call like: `var nodeGraph = NodeGraph.LoadGraphFromXml(xmlDoc, nodeFactory, out elementResolver);`

File: src/DynamoCore/Core/CustomNodeManager.cs [hunk 9]
```diff
@@ -1022,7 +1025,7 @@ public bool TryGetNodeInfo(string name, out CustomNodeInfo info)
                     Enumerable.Empty<NoteModel>(),
                     0,
                     0,
-                    newId, string.Empty);
+                    newId, currentWorkspace.ElementResolver, string.Empty);
```
Comment (aparajit-pratap) [inline src/DynamoCore/Core/CustomNodeManager.cs]: Again in this case, where a new `CustomNodeWorkspaceModel` is being created by collapsing nodes in an existing workspace, the `ElementResolver` is simply assigned to that of the parent workspace.

File: src/DynamoCore/Models/CustomNodeWorkspaceModel.cs (hunk 2 above)
Comment (aparajit-pratap) [inline src/DynamoCore/Models/CustomNodeWorkspaceModel.cs]: @Benglin as a response to this, please see my comments above and let me know if they make sense.

File: src/DynamoCore/Search/SearchDictionary.cs (hunk 3 above)
Comment (aparajit-pratap) [inline src/DynamoCore/Search/SearchDictionary.cs]: The 2 branches got mixed up. Bringing this code back in.

**Comment by aparajit-pratap** [comment]
@Benglin I have made the changes per your review. Please take a look again when you get the chance. Thanks.

**Comment by Benglin** [comment]
LGTM! Feel free to merge when you have sufficient time to address unit test failures. Thanks!
